core/.js/src/main/scala/io/udash/utils/FileService.scala

```diff
@@ -0,0 +1,172 @@
+package io.udash.utils
+
+import java.io.{IOException, InputStream}
+
+import org.scalajs.dom._
+import org.scalajs.dom.html.Anchor
+import org.scalajs.dom.raw.Blob
+import scalatags.JsDom
+
+import scala.scalajs.js
+import scala.concurrent.{Future, Promise}
+import scala.scalajs.js.annotation.JSGlobal
+import scala.scalajs.js.typedarray.ArrayBuffer
+import scala.util.Try
+
+@js.native
+@JSGlobal
+sealed class FileReaderSync() extends js.Object {
+  def readAsArrayBuffer(blob: Blob): ArrayBuffer = js.native
+}
+
+sealed class FileBufferedInputStream(file: File, bufferSize: Int) extends InputStream {
+  val fileReaderSync = new FileReaderSync()
+
+  var filePos: Int = 0
+  var pos: Int = 0
+
+  var buffer: Array[Byte] = Array.empty
+
+  override def available(): Int = {
+    val res = file.size.toInt - filePos
+    if (res < 0) 0 else res
+  }
+
+  override def read(): Int = {
+    if (pos >= buffer.length) {
+      import js.typedarray._
+
+      if (filePos >= file.size)
+        return -1
+
+      val len = math.min(filePos + bufferSize, file.size.toInt)
+      val slice = file.slice(filePos, len)
+      buffer = new Int8Array(fileReaderSync.readAsArrayBuffer(slice)).toArray
+      filePos += buffer.length
+      pos = 0
+    }
+    val r = buffer(pos).toInt & 0xff
+    pos += 1
+    r
+  }
+
+  override def close(): Unit = filePos = file.size.toInt
+
+  override def markSupported(): Boolean = true
+
+  var markPos: Option[Int] = None
+
+  override def mark(readlimit: Int): Unit =
+    markPos = Some(pos + filePos - buffer.length)
+
+  override def reset(): Unit = markPos match {
+    case Some(p) =>
+      filePos = p
+      pos = 0
+      buffer = Array.empty
+      markPos = None
+
+    case _ =>
+    // do nothing
+  }
+}
+
+object FileService {
+
+  final val OctetStreamType = "application/octet-stream"
+
+  /**
+   *  Converts specified bytes arrays to string that contains URL
+   *  that representing the array given in the parameter with specified mime-type.
+   *
+   *  Keep in mind that returned URL should be revoked via `org.scalajs.dom.revokeObjectURL(url)`.
+   */
+  def asURL(bytesArrays: Seq[Array[Byte]], mimeType: String): String = {
+    import js.typedarray._
+
+    val jsBytesArrays = js.Array[js.Any](bytesArrays.map(_.toTypedArray) :_ *)
+    val blob = new Blob(jsBytesArrays, BlobPropertyBag(mimeType))
+    URL.createObjectURL(blob)
```

Comment:
Maybe we should work on the return type and pack it in a class, which could have e.g. have `url` and `revoke` methods to notify the user of the workflow with these.

Reply:
What do you thing about brand new `CloseableUrl`?

Reply:
I think it's the right way to do it. Let's do `extends AbstractCase with AutoCloseable` to save some JS size.

Reply:
agreed